### ISSUE TITLE
Remove pipe-based git command line usage

### ISSUE THREAD

**Issue by rowanj** [rowanj/gitx#2]
GitX should use libgit2 & objective-git instead of running the git command line tools and parsing the output.

The pipe APIs used are deprecated, and they interact poorly with debug tools.

**Comment by rowanj** [comment]
Xcode 4.3 beta is distributed as a plain "Xcode.app" bundle, so git isn't even accessible from the command line by default with Xcode installed, by the look of it.  This bumps the priority of working on this, I'd like to have it done by the time Xcode 4.3 hits general release.

**Comment by rowanj** [comment]
To clarify the 'why' of this:
- The `git` binary that is found is not always the same one that the user will get on the command line
- Command responses are parsed from `stdout` and `stderr` of the binary.  Not all cases are handled, and it is fragile to any message format changes or localizations.
- It introduces a dependancy on an externally-installed program
- Any bugs due to the version/configuration of the user's `git` program will be difficult to diagnose or reproduce
- We want to be able to run out-of-the-box with current versions of Xcode, and it seems 4.3 is not installing command-line developer tools anymore

**Comment by rowanj** [comment]
We're going to be reliant on command-line git to some extent at least until libgit2 implements the git@ (ssh) transport.

Still, that shouldn't stop us from converting to libgit2 for local operations at least.

**Comment by rowanj** [comment]
n.b. GitX now looks for `git` as `xcrun git` as well, for the case where Xcode is installed but the Xcode command-line tools aren't.

**Comment by jtbandes** [comment]
libgit2 is currently missing some features: https://github.com/libgit2/libgit2/issues/1730

**Comment by rowanj** [comment]
@jtbandes true, but it does support pretty much everything that GitX has an interface for; and those things are coming.

I've merged a big hunk of commits to master for this recently, and would love any feedback people can add; esp. regarding loading performance or missing commits/branches.

The sort order for commits has also changed, I think for the better; but discussion is welcome.

**Comment by rowanj** [comment]
Problem repo in #183 tested with acceptable performance.

**Comment by sdegutis** [comment]
So apparently C is the new Unix? :cry:

**Comment by capnslipp** [comment]
@sdegutis Considering the shared authors between the two, I'd say C is the _old_ Unix.

**Comment by rowanj** [comment]
@sdegutis If you're asking for the justification for these changes:
- Parsing the command-line output is fragile; formats change, sort orders change, date representations are always a PITA. Profiling tools inject output into the stream, customised distributions, password agents, server messages and the like all combine to cause problems that can't be anticipated or diagnosed on test machines.
- It's slow; and can't be meaningfully parallelised.
- It introduces external dependencies beyond my control; bundling a version of Git introduces more problems when it interacts with the .gitconfig and repository format of the user's other tools.
- Users commonly have multiple `git` versions installed (i.e. Xcode, homebrew)
- Running shell commands from within the context of another program is a security risk
- OS X sandboxing makes this unreliable, too; and sandboxing is a Very Good Idea for applications that by requirement access remote resources and run code in the user's security context
